### PR TITLE
executor: Improve the performance of `WindowExec` by using sliding window

### PR DESCRIPTION
**Pull request by mmyj**
### What problem does this PR solve? 
Improve the performance of `WindowExec` by using sliding window or segment tree #12967

### What is changed and how it works?
1 add a interface named SlidingWindowAggFunc for evaluating the window aggregate functions using sliding window
2 modify the rowFrameWindowProcessor.appendResult2Chunk to use sliding window

### Check List 

Tests 
 - No code

Code changes

 - Has exported function/method change
 - Has interface methods change

### Benchmark
```
windowFuncs := []string{
	ast.AggFuncCount,
}
rows := []int{1000, 100000}
ndvs := []int{10, 1000}
frames := []*core.WindowFrame{
	{Type: ast.Rows, Start: &core.FrameBound{Type: ast.Preceding, Num: 10}, End: &core.FrameBound{Type: ast.Following, Num: 10}},
	{Type: ast.Rows, Start: &core.FrameBound{Type: ast.Preceding, Num: 100}, End: &core.FrameBound{Type: ast.Following, Num: 100}},
}
```

before
```
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)-8         	    4304	    249372 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)#01-8      	    1572	    760644 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)-8       	    4851	    230780 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)#01-8    	    5337	    232417 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)-8       	      51	  22412645 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)#01-8    	       8	 139406682 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)-8     	      52	  22229838 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)#01-8  	      15	  74082167 ns/op
```

now
```
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)-8         	   13956	     85242 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)#01-8      	   16231	     76437 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)-8       	    7408	    165960 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)#01-8    	    7059	    163368 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)-8       	     181	   6624173 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)#01-8    	     177	   6605687 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)-8     	     183	   6427620 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)#01-8  	     206	   5707492 ns/op
```

benchstat
```
name                                                                       old time/op  new time/op  delta
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)-8          447µs ± 1%   127µs ± 3%  -71.68%  (p=0.000 n=10+10)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)#01-8      1.82ms ± 1%  0.11ms ± 3%  -93.99%  (p=0.000 n=10+10)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)-8        265µs ± 1%   237µs ± 1%  -10.58%  (p=0.000 n=10+9)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)#01-8     266µs ± 1%   238µs ± 1%  -10.48%  (p=0.000 n=10+10)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)-8       44.8ms ± 1%  10.6ms ± 2%  -76.24%  (p=0.000 n=10+10)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)#01-8     350ms ± 3%    11ms ± 1%  -96.88%  (p=0.000 n=9+10)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)-8     43.0ms ± 1%  10.6ms ± 1%  -75.35%  (p=0.000 n=9+10)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)#01-8   181ms ± 1%     9ms ± 1%  -94.81%  (p=0.000 n=10+9)
```

### CODE REVIEW

File: executor/aggfuncs/aggfuncs.go [hunk 1]
```diff
@@ -138,3 +138,8 @@ type baseAggFunc struct {
 func (*baseAggFunc) MergePartialResult(sctx sessionctx.Context, src, dst PartialResult) error {
 	return nil
 }
+
+// SlidingWindowAggFunc is the interface to evaluate the aggregate functions using sliding window.
+type SlidingWindowAggFunc interface {
+	Slide(sctx sessionctx.Context, rows []chunk.Row, lastStart, lastEnd uint64, shiftStart, shiftEnd uint64, pr PartialResult) error
```
Comment (SunRunAway) [inline executor/aggfuncs/aggfuncs.go]: Please add comment for this method, it's better to describe each meaning of the arguments.

Comment (mmyj) [reply executor/aggfuncs/aggfuncs.go]: i will fix it.

**Comment by mmyj** [comment]
@SunRunAway I Implemented it without using interface like this https://github.com/mmyj/tidb/blob/33f969a9fe24cb6ce5640e3e2a9c2e700d1cb67e/executor/aggfuncs/aggfuncs.go#L127

and faster than before, the following are the benchmark result
```
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)-8         	   13826	     92467 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)#01-8      	   15820	     77616 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)-8       	    7863	    146567 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)#01-8    	    8036	    146462 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)-8       	     181	   6599287 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)#01-8    	     176	   6726682 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)-8     	     181	   6598735 ns/op
BenchmarkWindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)#01-8  	     206	   5799321 ns/op
```

File: executor/aggfuncs/func_count.go [hunk 2]
```diff
@@ -80,6 +105,31 @@ func (e *countOriginal4Real) UpdatePartialResult(sctx sessionctx.Context, rowsIn
 	return nil
 }
 
+func (e *countOriginal4Real) Slide(sctx sessionctx.Context, rows []chunk.Row, lastStart, lastEnd uint64, shiftStart, shiftEnd uint64, pr PartialResult) error {
+	p := (*partialResult4Count)(pr)
+	for i := uint64(0); i < shiftStart; i++ {
+		_, isNull, err := e.args[0].EvalInt(sctx, rows[lastStart+i])
```
Comment (SunRunAway) [inline executor/aggfuncs/func_count.go]: We'd better call `EvalReal` here.

File: executor/aggfuncs/func_count.go [hunk 3]
```diff
@@ -80,6 +105,31 @@ func (e *countOriginal4Real) UpdatePartialResult(sctx sessionctx.Context, rowsIn
...
+		if err != nil {
+			return err
+		}
+		if isNull {
+			continue
+		}
+		*p--
+	}
+	for i := uint64(0); i < shiftEnd; i++ {
+		_, isNull, err := e.args[0].EvalInt(sctx, rows[lastEnd+i])
```
Comment (SunRunAway) [inline executor/aggfuncs/func_count.go]: ditto.

**Comment by mmyj** [comment]
bechstat result
```
name \ time/op                                                             before.txt   slidingWindow.txt  without-interface.txt
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)-8          249µs ± 0%         124µs ± 0%              92µs ± 0%
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)#01-8       761µs ± 0%         113µs ± 0%              78µs ± 0%
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)-8        231µs ± 0%         227µs ± 0%             147µs ± 0%
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)#01-8     232µs ± 0%         227µs ± 0%             146µs ± 0%
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)-8       22.4ms ± 0%         9.7ms ± 0%             6.6ms ± 0%
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)#01-8     139ms ± 0%          10ms ± 0%               7ms ± 0%
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)-8     22.2ms ± 0%         9.7ms ± 0%             6.6ms ± 0%
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)#01-8  74.1ms ± 0%         8.9ms ± 0%             5.8ms ± 0%
```

**Comment by mmyj** [comment]
benchstat result
```
name                                                                       old time/op  new time/op  delta
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)-8          249µs ± 0%    85µs ± 0%   ~     (p=1.000 n=1+1)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:1000)#01-8       761µs ± 0%    76µs ± 0%   ~     (p=1.000 n=1+1)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)-8        231µs ± 0%   166µs ± 0%   ~     (p=1.000 n=1+1)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:1000)#01-8     232µs ± 0%   163µs ± 0%   ~     (p=1.000 n=1+1)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)-8       22.4ms ± 0%   6.6ms ± 0%   ~     (p=1.000 n=1+1)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:10,_rows:100000)#01-8     139ms ± 0%     7ms ± 0%   ~     (p=1.000 n=1+1)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)-8     22.2ms ± 0%   6.4ms ± 0%   ~     (p=1.000 n=1+1)
WindowFunctionsWithSlidingWindow/(func:count,_ndv:1000,_rows:100000)#01-8  74.1ms ± 0%   5.7ms ± 0%   ~     (p=1.000 n=1+1)
```

File: executor/window.go [hunk 4]
```diff
@@ -289,34 +289,63 @@ func (p *rowFrameWindowProcessor) consumeGroupRows(ctx sessionctx.Context, rows
...
+		shiftStart = start - lastStart
+		shiftEnd = end - lastEnd
 		for i, windowFunc := range p.windowFuncs {
-			err := windowFunc.UpdatePartialResult(ctx, rows[start:end], p.partialResults[i])
+			slidingWindowAggFunc := slidingWindowAggFuncs[i]
+			if slidingWindowAggFunc != nil && initializedSlidingWindow {
+				err = slidingWindowAggFunc.Slide(ctx, rows, lastStart, lastEnd, shiftStart, shiftEnd, p.partialResults[i])
+			} else {
+				err = windowFunc.UpdatePartialResult(ctx, rows[start:end], p.partialResults[i])
+				initializedSlidingWindow = true
```
Comment (SunRunAway) [inline executor/window.go]: If we have multiple `p.windowFuncs`, the second `windowFunc` will not run into this path.

File: executor/benchmark_test.go [hunk 5]
```diff
@@ -455,13 +465,21 @@ func benchmarkWindowExecWithCase(b *testing.B, casTest *windowTestCase) {
 		rows:   casTest.rows,
 		ctx:    casTest.ctx,
 	})
+	if casTest.frame != nil {
+		switch casTest.frame.Type {
+		case ast.Rows:
+			cmpFuncs := []expression.CompareFunc{expression.GetCmpFunction(cols[1], cols[1])}
+			casTest.frame.Start.CmpFuncs = cmpFuncs
+			casTest.frame.End.CmpFuncs = cmpFuncs
```
Comment (SunRunAway) [inline executor/benchmark_test.go]: nit: It seems that we do not need `comFuncs` with row frame?

**Comment by SunRunAway** [comment]
@qw4990 @XuHuaiyu PTAL

**Comment by mmyj** [comment]
@qw4990 @XuHuaiyu PTAL

**Comment by qw4990** [comment]
@mmyj LGTM, please resolve these conflicts~

**Comment by SunRunAway** [comment]
It has some tiny issues with prepare statement, waiting @mmyj to fix.

**Comment by mmyj** [comment]
Don't merge, maybe It has some tiny issues with prepare statement.

**Pull request merged**